cubi_tk/isa_tab/annotate.py

```diff
@@ -0,0 +1,458 @@
+"""``cubi-tk isa-tab annotate``: add annotation to ISA-tab from CSV file."""
+
+import argparse
+import csv
+import io
+import pathlib
+import typing
+from warnings import warn
+
+import attr
+from altamisa.constants.table_headers import (
+    SAMPLE_NAME,
+    MATERIAL_NAME_HEADERS,
+    PROCESS_NAME_HEADERS,
+    PROTOCOL_REF,
+    DATA_FILE_HEADERS,
+)
+from altamisa.isatab import InvestigationWriter, AssayWriter, StudyWriter, Characteristics, Comment
+from altamisa.isatab.helpers import is_ontology_term_ref
+from logzero import logger
+
+from .. import isa_support
+from ..common import overwrite_helper
+
+
+@attr.s(frozen=False, auto_attribs=True)
+class Config:
+    verbose: bool
+    config: str
+    force_update: bool
+    sodar_server_url: str
+    sodar_api_token: str = attr.ib(repr=lambda value: "***")  # type: ignore
+    no_warnings: bool
+    yes: bool
+    dry_run: bool
+    show_diff: bool
+    show_diff_side_by_side: bool
+    # batch_no: str
+    input_investigation_file: str
+    input_annotation_file: str
+    target_study: str
+    target_assay: str
+
+
+class SheetUpdateVisitor(isa_support.IsaNodeVisitor):
+    """IsaNodeVisitor that updates the ISA sample sheet as we walk along it."""
+
+    def __init__(self, annotation_map, header_map, overwrite, target_study, target_assay):
+        #: Mapping from normalized donor name to Donor instance.
+        self.annotation_map = annotation_map
+        self.header_map = header_map
+        self.overwrite = overwrite
+        #: The source names seen so far.
+        self.seen_source_names = set()
+        #: Study and assay selected for annotation update
+        self.target_study = target_study
+        self.target_assay = target_assay
+
+    def on_visit_material(self, material, node_path, study=None, assay=None):
+        super().on_visit_material(material, node_path, study, assay)
+
+        # Update material node only if part of targeted study or assay
+        if study.file.name.endswith(self.target_study) and not (
+            assay and not assay.file.name.endswith(self.target_assay)
+        ):
+            if material.type in DATA_FILE_HEADERS:
+                return self._update_comments(material)
+            else:  # Normal Materials
+                return self._update_characteristics(material, assay)
+
+        return None
+
+    def _has_content(self, value):
+        if is_ontology_term_ref(value):
+            return value.name or value.accession or value.ontology_name
+        else:
+            return value
+
+    def _update_characteristics(self, material, assay):
```

Comment:
This is complex method and definitely would benefit from more comments. Just by reading the code I cannot easily understand what are the cases in the if-else statements. particularly, the different in these returns:
```
 if assay and material.type == SAMPLE_NAME:
   return material

 ...

else:
   return material
```

Further, can this method be further divided? It seems like it dealing with different parts: `elif` block for comments, and another for header.

Reply:
I second this, please split each branch into its own method with its own docstring.